File: 8_atoi/myAtoi.go
var digits = map[byte]int{
	0x30: 0,
	0x31: 1,
	0x32: 2,
	0x33: 3,
	0x34: 4,
	0x35: 5,
	0x36: 6,
	0x37: 7,
	0x38: 8,
	0x39: 9,
}

func myAtoi(str string) int {
	res, sign, len, idx := 0, 1, len(str), 0

	// Skip leading spaces
	for idx < len && (str[idx] == ' ' || str[idx] == '\t') {
		idx++
	}

	if idx == len {
		return 0
	}

	// +/- Sign
	if str[idx] == '+' {
		sign = 1
		idx++
	} else if str[idx] == '-' {
		sign = -1
		idx++
	}

	// Digits: 0x30 = '0', 0x31 = '1', ... 0x39 = '9'
	for idx < len && str[idx] >= 0x30 && str[idx] <= 0x39 {
		res = res*10 + digits[str[idx]]
		if sign*res > math.MaxInt32 {
			return math.MaxInt32
		}

		if sign*res < math.MinInt32 {
			return math.MinInt32
		}

		idx++
	}

	return res * sign
}

File: 8_atoi/myAtoi.js
var myAtoi = function (s) {
  let i = 0,
    n = s.length,
    sign = 1,
    result = 0;

  while (i < n && s[i] === " ") i++;

  if (i < n && (s[i] === "+" || s[i] === "-")) {
    sign = s[i] === "-" ? -1 : 1;
    i++;
  }

  while (i < n && s[i] >= "0" && s[i] <= "9") {
    result = result * 10 + (s[i].charCodeAt(0) - "0".charCodeAt(0));
    if (result * sign > 2 ** 31 - 1) return 2 ** 31 - 1;
    if (result * sign < -(2 ** 31)) return -(2 ** 31);
    i++;
  }

  return result * sign;
};
